### PR TITLE
perf(Page): improve performance

### PR DESCRIPTION
**Pull request by zhangzhonghe**
### This is a ...
- [ ] New feature
- [x] Improvement
- [ ] Bug fix
- [ ] Others

### Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Improve page rendering performance and support page keep-alive functionality     |
| 🇨🇳 Chinese |     提升页面的渲染性能，和支持页面的 keep-alive 功能     |

### Checklists
- [x] All changes have been self-tested and work as expected
- [x] Test cases are updated/provided or not needed
- [x] Doc is updated/provided or not needed
- [x] Component demo is updated/provided or not needed
- [x] Changelog is provided or not needed
- [x] Request a code review if it is necessary

### CODE REVIEW

**Comment by Aminzarbani** [comment]
> 
> 
> ### This is a ...
> - [ ] New feature
> - [x] Improvement
> - [ ] Bug fix
> - [ ] Others
> 
> ### Motivation
> 
> 
> ### Description 
> 
> 
> ### Related issues
> 
> ### Showcase
> 
> 
> ### Changelog
> 
> | Language   | Changelog |
> | ---------- | --------- |
> | 🇺🇸 English |      Improve page rendering performance     |
> | 🇨🇳 Chinese |     提升页面的渲染性能      |
> 
> ### Checklists
> - [ ] All changes have been self-tested and work as expected
> - [ ] Test cases are updated/provided or not needed
> - [ ] Doc is updated/provided or not needed
> - [ ] Component demo is updated/provided or not needed
> - [ ] Changelog is provided or not needed
> - [ ] Request a code review if it is necessary
>

**Comment by Aminzarbani** [comment]
> 
> 
> ### This is a ...
> - [ ] New feature
> - [x] Improvement
> - [ ] Bug fix
> - [ ] Others
> 
> ### Motivation
> 
> 
> ### Description 
> 
> 
> ### Related issues
> 
> ### Showcase
> 
> 
> ### Changelog
> 
> | Language   | Changelog |
> | ---------- | --------- |
> | 🇺🇸 English |      Improve page rendering performance     |
> | 🇨🇳 Chinese |     提升页面的渲染性能      |
> 
> ### Checklists
> - [ ] All changes have been self-tested and work as expected
> - [ ] Test cases are updated/provided or not needed
> - [ ] Doc is updated/provided or not needed
> - [ ] Component demo is updated/provided or not needed
> - [ ] Changelog is provided or not needed
> - [ ] Request a code review if it is necessary
>

File: packages/core/client/src/acl/ACLProvider.tsx [hunk 1]
```diff
@@ -172,18 +182,17 @@ const getIgnoreScope = (options: any = {}) => {
 
 const useAllowedActions = () => {
   const service = useResourceActionContext();
-  const result = useBlockRequestContext();
-  return result?.allowedActions ?? service?.data?.meta?.allowedActions;
+  return service?.data?.meta?.allowedActions;
```
Comment (chenos) [inline packages/core/client/src/acl/ACLProvider.tsx:185]: 去掉了 useBlockRequestContext 是有问题的，可能对权限判断产生影响，这部分没有 e2e 覆盖

File: packages/core/client/src/appInfo/CurrentAppInfoProvider.tsx [hunk 2]
```diff
@@ -27,12 +26,9 @@ export const useCurrentAppInfo = () => {
   }>(CurrentAppInfoContext);
 };
 export const CurrentAppInfoProvider = (props) => {
-  const { render } = useAppSpin();
   const result = useRequest({
     url: 'app:getInfo',
   });
-  if (result.loading) {
-    return render();
-  }
```
Comment (chenos) [inline packages/core/client/src/appInfo/CurrentAppInfoProvider.tsx:36]: 为啥把这个去掉了？

File: packages/core/client/src/block-provider/BlockProvider.tsx [hunk 3]
```diff
@@ -97,34 +95,25 @@ export const MaybeCollectionProvider = (props) => {
...
+      resource,
+      __parent,
+      updateAssociationValues: props?.updateAssociationValues || [],
+    };
+  }, [__parent, field, props, resource, service]);
+
   return (
-    <BlockRequestContext_deprecated.Provider
-      value={{
-        allowedActions,
```
Comment (chenos) [inline packages/core/client/src/block-provider/BlockProvider.tsx:118]: 这里改了之后，权限相关的都出问题了

File: packages/core/client/src/block-provider/FormBlockProvider.tsx [hunk 4]
```diff
@@ -89,10 +87,6 @@ const InternalFormBlockProvider = (props) => {
     updateAssociationValues,
   ]);
 
-  if (service.loading && Object.keys(form?.initialValues || {})?.length === 0 && action) {
```
Comment (chenos) [inline packages/core/client/src/block-provider/FormBlockProvider.tsx:92]: 这个去掉也是有风险的，可能会导致数据表值有问题

File: packages/core/client/src/block-provider/TableBlockProvider.tsx [hunk 5]
```diff
@@ -61,7 +72,7 @@ const InternalTableBlockProvider = (props: Props) => {
     childrenColumnName,
     expandFlag: propsExpandFlag = false,
     fieldNames,
-    ...others
```
Comment (chenos) [inline packages/core/client/src/block-provider/TableBlockProvider.tsx:64]: 这个 ...others 是一堆参数，改变之后就只有 collection 了，有潜在问题

File: packages/core/client/src/block-provider/hooks/index.ts [hunk 6]
```diff
@@ -438,15 +439,15 @@ export const updateFilterTargets = (fieldSchema, targets: FilterTarget['targets'
...
+  const cm = useCollectionManager();
   const { getOperators } = useOperators();
   const fieldSchema = useFieldSchema();
   const { name } = useCollection();
   const { targets = [], uid } = useMemo(() => findFilterTargets(fieldSchema), [fieldSchema]);
 
   const getFilterFromCurrentForm = useCallback(() => {
-    return removeNullCondition(transformToFilter(form.values, getOperators(), getCollectionJoinField, name));
-  }, [form.values, getCollectionJoinField, getOperators, name]);
+    return removeNullCondition(transformToFilter(form.values, getOperators(), cm.getCollectionField.bind(cm), name));
```
Comment (chenos) [inline packages/core/client/src/block-provider/hooks/index.ts:449]: 这个也有风险，需要测试继承表里的情况

File: packages/core/client/src/block-provider/hooks/index.ts [hunk 7]
```diff
@@ -1296,12 +1301,12 @@ export const useAssociationFilterBlockProps = () => {
...
   const { props: blockProps } = useBlockRequestContext();
   const headers = useDataSourceHeaders(blockProps?.dataSource);
-  const cm = useCollectionManager_deprecated();
+  const cm = useCollectionManager();
   const { filter, parseVariableLoading } = useParsedFilter({ filterOption: field.componentProps?.params?.filter });
 
   let list, handleSearchInput, params, run, data, valueKey, labelKey, filterKey;
 
-  valueKey = collectionField?.target ? cm.getCollection(collectionField.target)?.getPrimaryKey() : 'id';
+  valueKey = collectionField?.target ? cm?.getCollection(collectionField.target)?.getPrimaryKey() : 'id';
```
Comment (chenos) [inline packages/core/client/src/block-provider/hooks/index.ts:1309]: 这一行也有问题，按理 cm 不会是 undefined，这里整段代码可以做个封装

File: packages/core/client/src/collection-manager/CollectionHistoryProvider.tsx [hunk 8]
```diff
@@ -65,16 +61,12 @@ export const CollectionHistoryProvider: React.FC = (props) => {
...
-    return render();
-  }
-
   return <CollectionHistoryContext.Provider value={value}>{props.children}</CollectionHistoryContext.Provider>;
 };
 
 export const useHistoryCollectionsByNames = (collectionNames: string[]) => {
   const { historyCollections } = useContext(CollectionHistoryContext);
-  return historyCollections.filter((i) => collectionNames.includes(i.name));
+  return historyCollections?.filter((i) => collectionNames.includes(i.name)) || [];
```
Comment (chenos) [inline packages/core/client/src/collection-manager/CollectionHistoryProvider.tsx:69]: History 的废弃了，可以考虑整个删掉

File: packages/core/client/src/schema-component/antd/grid/Grid.tsx [hunk 9]
```diff
@@ -427,16 +421,6 @@ Grid.Row = observer(
       };
     }, [fieldSchema, cols]);
 
-    const mapProperties = useCallback(
```
Comment (chenos) [inline packages/core/client/src/schema-component/antd/grid/Grid.tsx:430]: 这段删了会导致默认值 BUG

File: packages/core/client/src/data-source/collection-field/CollectionFieldProvider.tsx [hunk 10]
```diff
@@ -22,23 +22,25 @@ export type CollectionFieldProviderProps = {
   name?: SchemaKey;
   children?: ReactNode;
   allowNull?: boolean;
+  schema?: Schema;
 };
 
 export const CollectionFieldProvider: FC<CollectionFieldProviderProps> = (props) => {
-  const { name, children, allowNull } = props;
+  const { name, children, allowNull, schema } = props;
```
Comment (chenos) [inline packages/core/client/src/data-source/collection-field/CollectionFieldProvider.tsx:29]: 多了 schema 的有待考虑

File: packages/core/client/src/data-source/data-block/DataBlockRequestProvider.tsx [hunk 11]
```diff
@@ -163,11 +175,38 @@ export const BlockRequestProvider: FC = ({ children }) => {
...
+  return useMemo(() => (contextRef ? { ...contextRef.current, loading, data } : null), [contextRef, data, loading]);
+};
+
+/**
+ * Compared to `useDataBlockRequest`, the advantage of this hook is that it prevents unnecessary re-renders.
+ * For example, if you only need to use methods like `refresh` or `run`, it's recommended to use this hook,
+ * as it avoids component re-rendering when re-triggering requests.
+ * @returns
+ */
+export const useDataBlockRequestGetter = () => {
```
Comment (chenos) [inline packages/core/client/src/data-source/data-block/DataBlockRequestProvider.tsx:197]: 没有看懂这个的意图，这个是公共 api，看起来学习成本有点高

File: packages/core/client/src/modules/blocks/data-blocks/details-multi/setDataLoadingModeSettingsItem.tsx [hunk 12]
```diff
@@ -31,14 +30,14 @@ export function SetDataLoadingMode() {
...
-  const request = useDataBlockRequest();
+  const cm = useCollection();
+  const { getEnableFieldTree, getOnLoadData } = useCollectionState(cm?.name);
+  const { getDataBlockRequest } = useDataBlockRequestGetter();
 
   return (
     <SchemaSettingsModalItem
       title={t('Set data loading mode')}
-      scope={{ getEnableFieldTree, name, getOnLoadData }}
+      scope={{ getEnableFieldTree, name: cm?.name, getOnLoadData }}
```
Comment (chenos) [inline packages/core/client/src/modules/blocks/data-blocks/details-multi/setDataLoadingModeSettingsItem.tsx:40]: 这个不是 cm，而且不能为了不报错就 ?. 处理

Comment (zhangzhonghe) [reply packages/core/client/src/modules/blocks/data-blocks/details-multi/setDataLoadingModeSettingsItem.tsx:40]: collection 是有可能是空的，之前的也是可能是空的。之前不报错是因为原来那个上下文有个空对象的默认值。

命名，我改一下。

File: packages/core/client/src/schema-component/antd/block-item/BlockItem.style.tsx
Comment (NameNoQuality) [inline packages/core/client/src/schema-component/antd/block-item/BlockItem.style.tsx:1]: ~~○~~

**Comment by zhangzhonghe** [comment]
The e2e test errors in approval can be ignored for now. Other errors can pass when running locally, so it can be merged.

**Pull request merged**